UI/API changes to facilitate disabling features within spaces

legrego opened the pull request:
This PR introduces a UI to enable users to hide kibana features within specific spaces. Since this is available under a basic license, the functionality is not a security feature, but more of a user experience feature.

## Implementation
This adds a new `disabledFeatures` field to the `space` saved object type. This is a keyword array which stores the IDs of each feature that is disabled. This allows new features to be visible by default with a Basic license.

## Current UI
![image](https://user-images.githubusercontent.com/3493255/48485913-12e18a80-e7e8-11e8-8480-1cd95b39085f.png)


## (OUTDATED) Proposed UI:
![image](https://user-images.githubusercontent.com/3493255/48370695-0cd49800-e688-11e8-9944-40b09c61e862.png)



## (OUTDATED) WIP UI:
![image](https://user-images.githubusercontent.com/3493255/47660260-0bc24780-db6d-11e8-8b04-bb7e50c2dcb1.png)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/5394/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/5536/)

legrego commented:
@kobelb UI/UX aside, what are your thoughts on the general approach here? Happy to go over this on zoom, or after we figure out the user profile replacement.

legrego commented:
> The way that you're using the current uiExports to drive this is very interesting, I'd like to see how this plays out once userProfiles are fleshed out a bit more and we have a little more understanding for how we're going to make the privileges model extensible.

Yeah, I wouldn't be surprised if we end up with something entirely different in the end for that. It's also quite likely that the feature list we show in Spaces Management will be different from the list we show in Role Management. For example, we might not want to allow users to hide index pattern management with a basic license, because that would render the Space somewhat unusable for everyone, unless they hide the feature after setting up their dashboards and such. 🤷‍♂️

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/6828/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/7037/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/7062/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/7063/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/7714/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/7826/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/7841/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/7851/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/7881/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/7896/)

legrego commented:
@cchaos I apologize if we already talked about this, but do you have any thoughts on if or how we should allow for filtering the list of features in the screenshot above?

Adding the traditional table search bar doesn't seem like a great idea to me, because it ends up looking a lot like another input field in the form.


I was also thinking about rewriting this screen using `DescribedFormGroups` instead, which would give us some room to explain the new settings, and maybe help visually separate the name/avatar/description from the new settings. Do you think that would help, or would it just create too much noise?

cchaos commented:
Off the bat, I would almost start using the whole pattern of the edit role page where you use the descriptive form rows but also split up related content. So the first section would be the default form and then the next section would be the enable/disable features. Then I'd have that section collapsed by default since it's pushing the "Save" button far down the page and they may not care to always change these options.

That'll allow the table search bar to be in it's own section and hopefully not look like an input. Though, I'm not sure search is necessary for 13 items.

Could you also reverse the columns and use a switch instead of checkbox like this:

<img width="254" alt="screen shot 2018-11-08 at 09 37 40 am" src="https://user-images.githubusercontent.com/549577/48205324-08337b00-e33a-11e8-8ca2-06f36a9df1d2.png">

legrego commented:
Thanks @cchaos!

Here's what I have so far. I know the copy will have to change, but I wanted some placeholder text for the time being:

## Initial page load
![image](https://user-images.githubusercontent.com/3493255/48268961-704e9380-e404-11e8-92a9-2ac142442685.png)

## Expanded features section
![image](https://user-images.githubusercontent.com/3493255/48269240-169a9900-e405-11e8-80ff-6d65748a8652.png)




A couple of thoughts:
1) I'm reusing the same component from the Role Management screen to allow those regions to be collapsed. The titles feel larger to me here, even though they should be the same size. Do you think they're too big?
2) I wasn't sure what icon to use for the "Enabled Features" section. Do you have a better idea? Is this something that we should get a new icon for? I briefly considered the `logoElasticStack`, but these are features just within Kibana, and I thought that'd be misleading.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/8322/)

cchaos commented:
Yeah this is looking better.

1. I can't imagine why the titles would be bigger when using the same component. I also don't have a reference point right now to compare. I can always take a look and do a design review when you are ready.

2. I'd just get rid of the icons. They're not really necessary as all the settings apply to spaces and kibana. I would probably even get rid of the title and collapsibility from the first one entirely and just have the second one be collapsible (with no icon).

cchaos commented:
Oh and I talked to the design team and we can get rid of the "Yes/No" labels of the switches but be sure to pass `aria-label` or similar for accessibility.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/8346/)

legrego commented:
@gchaps What are your thoughts on the copy I have so far? I mostly added placeholder text, so I assume that it'll need quite a few changes. Happy to talk about this further if you'd like more context!

@cchaos thoughts on what I have at this point? If nothing jumps out, I then I think this is ready for a more in-depth design review. I'm planning on adding i18n support after we've finished design and copy reviews

![image](https://user-images.githubusercontent.com/3493255/48370754-38f01900-e688-11e8-96e1-920613e922a1.png)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/8455/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/8490/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/8559/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/8568/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/8596/)

cchaos commented:
@legrego I'm almost ready with a design PR but I have 2 questions:

1. I'm having an issue with the customizing the URL identifier. Why does it need an "edit" link to trigger edit-ability when no matter the edit-ability state, it always gets overridden when you change the space name?

<img src="https://d.pr/free/i/OQ2l4z+" />

2. I see the list of features only includes items available in a basic license. What about x-pack features?

<img width="887" alt="screen shot 2018-11-13 at 16 34 35 pm" src="https://user-images.githubusercontent.com/549577/48444476-09154400-e762-11e8-9218-59d522bee854.png">

legrego commented:
@cchaos 

1. I don't have any technical reasons why we need that edit link, so I'd be ok removing that if that simplifies things.

2. Adding x-pack features isn't part of this PR -- Brandon and I have about 4 PRs in flight right now, and once they all merge together, we'll be able to get x-pack features added here without making any changes to this screen.

cchaos commented on x-pack/plugins/spaces/public/views/management/_manage_spaces.scss:
```diff
@@ -1,3 +1,5 @@
+@import 'x-pack/public/components/collapsible_panel/collapsible_panel';
```
It's unnecessary to import other SASS files because all SASS files (including all in X-pack) get compiled and loaded on all pages, so importing again will duplicate the styles.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/8609/)

cchaos commented:
Can you use the `edit` button to allow the url to always be the custom one provided and no longer auto-create the url if the name changes? Maybe just change the text of the button to be `customize` and `reset` (which will reset to the auto-created version & read-only)?

legrego commented:
> Can you use the edit button to allow the url to always be the custom one provided and no longer auto-create the url if the name changes? Maybe just change the text of the button to be customize and reset (which will reset to the auto-created version & read-only)?

Yep, I think that's a good idea. I could have sworn it used to work that way, so either I dreamed that up, or I broke it at some point. I'll make that change either way, and adjust the link text too

cchaos commented:
PR4U: https://github.com/legrego/kibana/pull/12

legrego commented:
This looks so much better, thanks again @cchaos! I have the popover wired up, and I made your suggested update to change "edit" / "stop editing" => "customize" / "reset"

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/8697/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/263/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/266/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/417/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/423/)

gchaps commented:
Because this changes the visibility of the features, and not whether the feature is on or off, I thought we "enable" wasn't the right word.   How about something like this?

Change feature display    all features visible   [hide]

Control which features are visible in this space.  

This is not a security mechanism.  

And for the table labels, use "Feature" and "Show?"

gchaps commented:
Here are some additional edits for the top part of the page.

Create a space

---

Organize your saved objects into meaningful categories.

---

Customize your space 

Name your space and customize its avatar.

Note the URL identifier.  You cannot change it after you create the space.

---

The description appears on the Space selection screen.

legrego commented:
@gchaps Here's a screenshot with your recommendations applied:

![image](https://user-images.githubusercontent.com/3493255/49042541-6cf23080-f196-11e8-80bd-3da613d9c0d1.png)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1286/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1288/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1292/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1302/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1310/)

kobelb commented on x-pack/plugins/spaces/server/lib/toggle_ui_capabilities.ts:
```diff
@@ -5,30 +5,41 @@
 
-export async function toggleUiCapabilities(uiCapabilities: UICapabilities, activeSpace: Space) {
-  toggleDisabledFeatures(uiCapabilities, activeSpace);
+export function toggleUiCapabilities(
```
nit: can we change this to `toggleUICapabilities` to be "consistent" with elsewhere?

legrego replied:
we sure can!

gchaps commented:
@legrego just a few questions/tweaks to the text:

- Instead of "Create space" can the title be "Create **a** space"?
- Can "security mechanism" be "security feature" instead? (its easier to say)
- Change "Change feature display" to "**Control** feature display"

Also, should "Customize your space" and "Control feature display" be the same font size?  That would be a question for design.

legrego commented on x-pack/plugins/spaces/public/views/management/edit_space/customize_space/space_identifier.tsx:
```diff
@@ -79,74 +79,77 @@ class SpaceIdentifierUI extends Component<Props, State> {
           id="xpack.spaces.management.spaceIdentifier.kibanaURLForEngineeringIdentifierDescription"
-          defaultMessage="If the identifier is {engineeringIdentifier}, the Kibana URL is{nextLine}
-          {engineeringKibanaUrl}."
+          defaultMessage="Example: https://my-kibana.example{spaceIdentifier}/app/kibana."
```
Credit goes to @cchaos for this 🥇

legrego commented on x-pack/plugins/spaces/public/views/management/edit_space/customize_space/space_identifier.tsx:
```diff
@@ -79,74 +79,77 @@ class SpaceIdentifierUI extends Component<Props, State> {
         <FormattedMessage
           id="xpack.spaces.management.spaceIdentifier.urlIdentifierLabel"
-          defaultMessage="URL identifier"
+          defaultMessage="URL identifier "
```
Yep, this helps space things correctly now that this is a `FormattedMessage`

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1524/)

legrego commented:
@gchaps I had a second thought on changing "This is not a security mechanism" -> "This is not a security feature":

Are we overloading the term "feature" here by making this change? We'd be asking users to control which Kibana features are shown/hidden, using a feature of spaces.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1530/)

legrego commented on x-pack/plugins/spaces/server/lib/migrations/migrate_6x.ts:
```diff
@@ -0,0 +1,12 @@
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+export function migrateToKibana660(doc: Record<string, any>) {
```
We discussed, and agreed this is fine as-is given how migrations currently work

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1537/)

legrego commented:
@kobelb This is ready for another review when you get a chance. To get CI to run, I had to rebase against master, which merged in the "Delete privileges" pr that we merged yesterday. I had a conflict on a test file, so I'd appreciate a second look at [this file specifically](https://github.com/elastic/kibana/pull/24235/files#diff-44dc018370a909d69de3231af10bda6d) to make sure I didn't mess up any scenarios.

cchaos commented:
@legrego , I would do the opposite and increase the size of "Customize your space" instead of decreasing the size of "Change feature display"

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1551/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1724/)

legrego commented:
@AlonaNadler thanks for catching the issue with disabled features staying disabled when they shouldn't be. I updated this with a fix.

kobelb commented:
Thanks for fixing those breaking tests I introduced with the change to being able to delete privileges

legrego commented:
@gchaps @AlonaNadler I know that we haven't addressed all feedback yet, but I am going to merge this into our feature branch to unblock work on other components of granular application privileges.

Once we all agree on a set of changes, then we can create an issue outlining them for us to address.
